// @vercel/build-utils/dist/workspaces/get-workspace-package-paths.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWorkspacePackagePaths = void 0;
const path_1 = __importDefault(require("path"));
const js_yaml_1 = __importDefault(require("js-yaml"));
const glob_1 = __importDefault(require("glob"));
const get_glob_fs_1 = require("../fs/get-glob-fs");
const normalize_path_1 = require("../fs/normalize-path");
const posixPath = path_1.default.posix;
async function getWorkspacePackagePaths({ fs, workspace, }) {
    const { type, rootPath } = workspace;
    const workspaceFs = fs.chdir(rootPath);
    let results = [];
    switch (type) {
        case 'yarn':
        case 'npm':
            results = await getPackageJsonWorkspacePackagePaths({ fs: workspaceFs });
            break;
        case 'pnpm':
            results = await getPnpmWorkspacePackagePaths({ fs: workspaceFs });
            break;
        default:
            throw new Error(`Unknown workspace implementation: ${type}`);
    }
    return results.map(packagePath => {
        return posixPath.join(rootPath, posixPath.dirname(packagePath));
    });
}
exports.getWorkspacePackagePaths = getWorkspacePackagePaths;
async function getPackagePaths(packages, fs) {
    return (await Promise.all(packages.map(packageGlob => new Promise((resolve, reject) => {
        glob_1.default(normalize_path_1.normalizePath(posixPath.join(packageGlob, 'package.json')), {
            cwd: '/',
            fs: get_glob_fs_1.getGlobFs(fs),
        }, (err, matches) => {
            if (err)
                reject(err);
            else
                resolve(matches);
        });
    })))).flat();
}
async function getPackageJsonWorkspacePackagePaths({ fs, }) {
    const packageJsonAsBuffer = await fs.readFile('package.json');
    const { workspaces } = JSON.parse(packageJsonAsBuffer.toString());
    let packages = [];
    if (Array.isArray(workspaces)) {
        packages = workspaces;
    }
    else {
        packages = workspaces?.packages ?? [];
    }
    return getPackagePaths(packages, fs);
}
async function getPnpmWorkspacePackagePaths({ fs, }) {
    const pnpmWorkspaceAsBuffer = await fs.readFile('pnpm-workspace.yaml');
    const { packages = [] } = js_yaml_1.default.load(pnpmWorkspaceAsBuffer.toString());
    return getPackagePaths(packages, fs);
}
